### PR TITLE
systemd: Fix -Dpid arg passing to systemd usage service

### LINKED ISSUE #3203: usage server broken in 4.12.0.0
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME
Usage server
~~~

~~~

##### CLOUDSTACK VERSION

version 4.12, downloaded from official repo 

##### CONFIGURATION
vanilla install

##### OS / ENVIRONMENT
Centos 7

##### SUMMARY
Usage server not starting jobs and ignore any db.usage settings

[root@ref-trl-1013-k-M7-apanic-mgmt1 usage]# grep -i usage /etc/cloudstack/usage/db.properties | grep invalid
db.usage.name=invalid_database_name
db.usage.port=invalid_port
db.usage.username=invalid_username
db.usage.password=invalid_password!
[root@ref-trl-1013-k-M7-apanic-mgmt1 usage]# tail -n 5 /var/log/cloudstack/usage/usage.log
2019-03-07 22:40:36,738 INFO  [cloud.usage.UsageManagerImpl] (main:null) (logid:) Implementation Version is 4.12.0.0
2019-03-07 22:40:37,569 DEBUG [cloud.usage.UsageManagerImpl] (main:null) (logid:) Usage stats aggregation time zone: GMT
2019-03-07 22:40:37,570 DEBUG [cloud.usage.UsageManagerImpl] (main:null) (logid:) Execution Time: Thu Mar 07 22:28:00 UTC 2019
2019-03-07 22:40:37,570 DEBUG [cloud.usage.UsageManagerImpl] (main:null) (logid:) Current Time: Thu Mar 07 22:40:37 UTC 2019
2019-03-07 22:40:37,575 INFO  [cloud.usage.UsageServer] (main:null) (logid:) UsageServer ready..

##### STEPS TO REPRODUCE
try to install usage server from http://download.cloudstack.org/centos7/4.12/
and force it to proceed with usage records, no luck
My info here in user mail list
http://mail-archives.apache.org/mod_mbox/cloudstack-users/201903.mbox/%3CVI1PR07MB4816C12D23D511AD0904D9D9E44C0%40VI1PR07MB4816.eurprd07.prod.outlook.com%3E

##### EXPECTED RESULTS

~~~
usage server need to be fixed. That is show stopper for the release
~~~

##### ACTUAL RESULTS

~~~

~~~

### PR DESCRIPTION
**Pull request by rhtyd**
This fixes regression introduced by refactoring PR #3163 where `-Dpid`
was incorrectly passed string `$$` instead of parent PID integer.

Fixes #3203

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

**Comment by rhtyd** [comment]
cc @andrijapanic  @PaulAngus @ustcweizhou

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

File: packaging/systemd/cloudstack-usage.service [hunk 1]
```diff
@@ -24,7 +24,8 @@ After=network.target network-online.target
 [Service]
 Type=simple
 EnvironmentFile=/etc/default/cloudstack-usage
-ExecStart=/usr/bin/java $JAVA_OPTS -cp $CLASSPATH $JAVA_CLASS
+Environment=JAVA_PID=$$
```
Comment (rhtyd) [inline packaging/systemd/cloudstack-usage.service]: @GabrielBrascher @andrijapanic moved the JAVA_PID here so users can't change/mess it around. The systemd variable referencing is limiting, the fix was to use the `${VAR}` syntax and not `$VAR` syntax. Can you believe it! https://www.freedesktop.org/software/systemd/man/systemd.service.html#Command%20lines

**Comment by rhtyd** [comment]
@GabrielBrascher fixed the issue, it was systemd variable referencing/syntax issue which per following docs should be `${VAR}` instead of `$VAR`:
https://www.freedesktop.org/software/systemd/man/systemd.service.html#Command%20lines

After my last commit, it's working again.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by andrijapanicsb** [comment]
LGTM

Tested and the job does kick in as expected.

**Comment by GabrielBrascher** [comment]
@rhtyd @andrijapanic are you testing it with CentOS? On Ubuntu I still get error. Now the exception is: `java.lang.NumberFormatException: For input string: "$$"`

**Comment by andrijapanicsb** [comment]
yes, Centos7
let me build Ubuntu lab - 18.04 sounds fine ?

**Comment by GabrielBrascher** [comment]
@rhtyd @andrijapanic nevermind, my mistake when updating the script at this time. Due to the /bin/sh -ec '' it works now. Thanks!

**Comment by andrijapanicsb** [comment]
Cool - let's LGTM this one and close if you are OK with it @GabrielBrascher  ?

**Comment by GabrielBrascher** [comment]
@rhtyd it looks good. Thanks!

Can you please also remove the lines that are silencing the exception?
https://github.com/apache/cloudstack/blob/1d05fead49f5c856257a741b07122f5633d2e359/usage/src/main/java/com/cloud/usage/UsageServer.java#L54

```
} catch (Exception e) {
    e.printStackTrace();
}
```

**Comment by rhtyd** [comment]
@GabrielBrascher okay I'll move that to logging tomorrow.
@blueorangutan test

**Comment by ustcweizhou** [comment]
tested ok in 4.11.2
LGTM

**Comment by GabrielBrascher** [comment]
All checks have passed (Travis, Jenkins, Trillian) passed and we have 3 LGTMs. I also tested and it worked. @rhtyd this PR is looking good.

@rhtyd just to keep track on the exception silenced; will you be able to add a commit addressing that exception silenced? Adding the JAVA_DEBUG is great, fixing that exception is also a great plus to help users on debugging.

Thanks for the work!

**Comment by rhtyd** [comment]
@GabrielBrascher sorry forgot about that amidst dayjob work. I can do that tomorrow, or if you'd prefer you nay push your changes to this PR's branch.

**Comment by GabrielBrascher** [comment]
No problem @rhtyd, thanks!

**Comment by rhtyd** [comment]
@GabrielBrascher done

**Comment by GabrielBrascher** [comment]
Thanks, @rhtyd! I am merging this PR.

**Comment by rhtyd** [comment]
@GabrielBrascher, I've forward merged it to master now, in case you want to cut the RC or address another issue. Thanks.

**Comment by GabrielBrascher** [comment]
Time to spin out RC5 then, thanks!

**Pull request merged**